### DISCUSSION TITLE
Why can't we have non React files inside our pages folder

### DISCUSSION THREAD

**Discussion by jacktan165** [vercel/next.js#12336] [Help] [answered]
Before Next.js, I used to have my tests and storybook files next to my page component, like so:

```
App.tsx
App.test.tsx
App.stories.tsx
```

Next.js disallows that, because when you try to run `next build`, it complains that `describe is not defined` etc. I'm going to assume it prefers us to put all our non React files outside the page directory in /tests, /stories.
But I don't want to.
Having all App related files side by side helps alot in maintainability, especially in a large project. Is there a way to do this instead?

**Comment by jacktan165** [comment, marked as answer]
Solved it. For anyone wanting to know what I did:

on `next.config.js`, add this:
```
    config.plugins.push(new webpack.IgnorePlugin(/test.tsx$/));
    config.plugins.push(new webpack.IgnorePlugin(/stories.tsx$/));
```

**Comment by timneutkens** [comment]
Hid the comment as this is not recommended and is likely to break between version changes.

The reasoning for this has been discusses in many issues, eg https://github.com/zeit/next.js/issues/4315

**Reply by rtrembecky** [reply]
Hi, the #4315 is about changing the `pages` directory, I didn't find information about allowing non-page files in the `pages` folder. However, I've seen a lot of your comments about minimizing the number of config options and possibly introducing more bugs, so is it similar in this case? I can imagine having a config option used to tell next.js to ignore `*.test.@(ts|tsx)` and `*.stories.@(ts|tsx)`.

So we are currently not able to put tests and stories next to the modules for all files (as outlined as the best structure e.g. here https://stackoverflow.com/a/62765143 and being the recommended one by jest), forcing us to have some not-so-clean hybrid structure. Is this the testing/storybook folder structure currently recommended by the next.js team? I know about this example but there are no component tests: https://github.com/arcatdmz/nextjs-with-jest-typescript

**Reply by airtonix** [reply]
yeah you should really be organising your files like this : 

```
pages/
  index.tsx
  other.tsx
  something/
    something/
      something/
        darkside.tsx
components/
  something/
    TheDarksidePage/
      index.ts
      TheDarksidePage.tsx
      TheDarksidePage.stories.tsx
      TheDarksidePageContainer.tsx

  other/
   OtherPage/
     OtherPage.tsx
     OtherPage.stories.tsx
     OtherPageContainer.tsx
     index.ts

  landing/
   LandingPage/
     index.ts
     LandingPage.tsx
     LandingPage.stories.tsx
     LandingPage.test.tsx
     LandingPageContainer.tsx

   SomeKindOfWelcomeBanner/
     index.ts
     SomeKindOfWelcomeBanner.tsx
```

`something/something/something/darkside.tsx` 

```tsx

import { TheDarksidePageContainer } from 'components/something/TheDarksidePage';

function TheDarksideRoute() {
 const router = useRouter()
 
 const handleNavigateToInductionClick = () => {
  router.push('/other')
 }

 return (
  <TheDarksidePageContainer onNavigateToInductionClick={handleNavigateToInductionClick} />
 );
}

export default TheDarksideRoute;

export function getStaticProps() {
  return {
    ...
  }
}
```


Then the following rules apply: 

- prop drilling is not evil or bad. you're being lazy. suck it up.
- only container files can import container files (use composition!!!)
- container files can import non container files
- `useRouter` is only allowed to be used in components directly in `pages/**` (we use context to make our link component available downstream).

**Comment by sldavidson** [comment]
In case anyone stumbles across this discussion: https://nextjs.org/docs/api-reference/next.config.js/custom-page-extensions#including-non-page-files-in-the-pages-directory
